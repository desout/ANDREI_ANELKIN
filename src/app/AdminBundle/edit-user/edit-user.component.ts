import {Component, Inject, OnInit} from '@angular/core';
import {UserDropDownService} from '../../user-dropdown.service';
import {FormBuilder, FormGroup, Validators} from '@angular/forms';
import {User} from '../../../../server/server/User';
import {UserService} from '../../user.service';
import {TranslateService} from '@ngx-translate/core';
import {Store} from '@ngrx/store';
import { CurUserState, UsersState} from '../../store';
import {isNameTaken} from '../../validators/user-name.validator';
import {isAvailableName} from '../../validators/user-name-available.validator';
import {DateValidator} from '../../validators/date.validator';
import {AddUser, DeleteUser, GetUsers, UpdateUser} from '../../store/user-store/actions/users.actions';
import {EditUsersComponent} from '../edit-users/edit-users.component';

@Component({
  selector: 'app-edit-user',
  templateUrl: './edit-user.component.html',
  styleUrls: ['./edit-user.component.scss']
})
export class EditUserComponent implements OnInit {
  userForm: FormGroup;
  getUserFromForm(): User {
    return {
      name:  this.userForm.get('name').value,
      dateOfBirth: this.userForm.get('dateOfBirth').value,
      dateOfFirstLogin: this.userForm.get('dateOfFirstLogin').value,
      dateNextNotification: this.userForm.get('dateNextNotification').value,
      information: this.userForm.get('information').value,
      role: ''
    };
  }
  constructor(public userService: UserService,
              public dropdownService: UserDropDownService,
              private formBuilder: FormBuilder,
              public translate: TranslateService,
              private usersStore: Store<UsersState>,
              @Inject(EditUsersComponent) private parent: EditUsersComponent) {}

  ngOnInit(): void {
      this.userForm = this.formBuilder.group({
        name: ['', Validators.compose([Validators.required, Validators.minLength(4)]),
          [isNameTaken.bind(this), isAvailableName(this.userService, '').bind(this)],
          'blur'],
        dateOfBirth: ['', Validators.compose([Validators.required, DateValidator])],
        dateOfFirstLogin: ['', Validators.compose([Validators.required, DateValidator])],
        dateNextNotification: ['', Validators.compose([Validators.required, DateValidator])],
        information: ['', Validators.compose([Validators.required, Validators.minLength(4)])]

      });
  }
  onSubmit() {
    const newUser: User = this.getUserFromForm();
    if (!this.dropdownService.selectedUser) {
      newUser.role = 'USER';
      this.usersStore.dispatch(new AddUser(newUser));
    } else {
      newUser.role = this.dropdownService.selectedUser.role;
      newUser.id = this.dropdownService.selectedUser.id;
      this.usersStore.dispatch( new UpdateUser(newUser));
    }
    this.parent.showEdit = false;
  }
  onClickDelete() {
    if (this.dropdownService.selectedUser.role !== 'ADMIN') {
      this.usersStore.dispatch(new DeleteUser(this.dropdownService.selectedUser.id));
      this.usersStore.dispatch(new GetUsers());
      this.userForm.reset();
      this.dropdownService.selectedUserHandle.next(-1);
    }
  }

  get name() { return this.userForm.get('name'); }
  get dateOfBirth() { return this.userForm.get('dateOfBirth'); }
  get dateOfFirstLogin() { return this.userForm.get('dateOfFirstLogin'); }
  get dateNextNotification() { return this.userForm.get('dateNextNotification'); }
  get information() { return this.userForm.get('information'); }

}
