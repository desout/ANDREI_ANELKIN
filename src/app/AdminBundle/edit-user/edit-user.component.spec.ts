import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { EditUserComponent } from './edit-user.component';
import {NO_ERRORS_SCHEMA} from '@angular/core';
import {TestModule} from '../../Test.module';

describe('EditUserComponent', () => {
  let component: EditUserComponent;
  let fixture: ComponentFixture<EditUserComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [],
      schemas: [NO_ERRORS_SCHEMA],
      imports: [TestModule],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(EditUserComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
