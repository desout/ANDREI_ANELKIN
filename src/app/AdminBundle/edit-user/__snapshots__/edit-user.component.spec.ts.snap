// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUserComponent should create 1`] = `
<div
  id="root0"
  ng-version="7.0.3"
>
  <br />
  <div
    class="container"
  >
    <form
      class="ng-untouched ng-pristine ng-invalid"
      ng-reflect-form="[object Object]"
      novalidate=""
    >
      <div
        class="form-group"
      >
        <label>
          INFO.USERNAME 
          <input
            class="form-control ng-untouched ng-pristine ng-invalid"
            formcontrolname="name"
            ng-reflect-name="name"
          />
        </label>
        
        
      </div>
      <div
        class="form-group"
      >
        <label>
          INFO.BIRTHDAY 
          <input
            class="form-control ng-untouched ng-pristine ng-invalid"
            formcontrolname="dateOfBirth"
            ng-reflect-name="dateOfBirth"
            ng-reflect-required=""
            required=""
          />
        </label>
        
      </div>
      <div
        class="form-group"
      >
        <label>
          INFO.DATEOFLOGIN 
          <input
            class="form-control ng-untouched ng-pristine ng-invalid"
            formcontrolname="dateOfFirstLogin"
            ng-reflect-name="dateOfFirstLogin"
            ng-reflect-required=""
            required=""
          />
        </label>
        
      </div>
      <div
        class="form-group"
      >
        <label>
          INFO.DATEOFNEXTNOTIF 
          <input
            class="form-control ng-untouched ng-pristine ng-invalid"
            formcontrolname="dateNextNotification"
            ng-reflect-name="dateNextNotification"
            ng-reflect-required=""
            required=""
          />
        </label>
        
      </div>
      <div
        class="form-group"
      >
        <label>
          INFO.INFORMATION 
          <input
            class="form-control ng-untouched ng-pristine ng-invalid"
            formcontrolname="information"
            ng-reflect-name="information"
            ng-reflect-required=""
            required=""
          />
        </label>
        
      </div>
      <button
        class="btn btn-default"
        disabled=""
        type="submit"
      >
        INFO.SUBMITBTN
      </button>
    </form>
  </div>
</div>
`;
